File: src/libtraci/TrafficLight.cpp
/****************************************************************************/
// Eclipse SUMO, Simulation of Urban MObility; see https://eclipse.org/sumo
// Copyright (C) 2017-2021 German Aerospace Center (DLR) and others.
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License 2.0 which is available at
// https://www.eclipse.org/legal/epl-2.0/
// This Source Code may also be made available under the following Secondary
// Licenses when the conditions for such availability set forth in the Eclipse
// Public License 2.0 are satisfied: GNU General Public License, version 2
// or later which is available at
// https://www.gnu.org/licenses/old-licenses/gpl-2.0-standalone.html
// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-or-later
/****************************************************************************/
/// @file    TrafficLight.cpp
/// @author  Daniel Krajzewicz
/// @author  Mario Krumnow
/// @author  Jakob Erdmann
/// @author  Michael Behrisch
/// @date    30.05.2012
///
// C++ TraCI client API implementation
/****************************************************************************/
#include <config.h>

#define LIBTRACI 1
#include <libsumo/StorageHelper.h>
#include <libsumo/TraCIConstants.h>
#include <libsumo/TrafficLight.h>
#include "Domain.h"

// TODO remove the following line once the implementation is mature
#ifdef _MSC_VER
#pragma warning(disable: 4100)
#else
#pragma GCC diagnostic ignored "-Wunused-parameter"
#endif

namespace libtraci {

typedef Domain<libsumo::CMD_GET_TL_VARIABLE, libsumo::CMD_SET_TL_VARIABLE> Dom;

// ===========================================================================
// static member definitions
// ===========================================================================
std::vector<std::string>
TrafficLight::getIDList() {
    return Dom::getStringVector(libsumo::TRACI_ID_LIST, "");
}


int
TrafficLight::getIDCount() {
    return Dom::getInt(libsumo::ID_COUNT, "");
}


std::string
TrafficLight::getRedYellowGreenState(const std::string& tlsID) {
    return Dom::getString(libsumo::TL_RED_YELLOW_GREEN_STATE, tlsID);
}


std::vector<libsumo::TraCILogic>
TrafficLight::getAllProgramLogics(const std::string& tlsID) {
    tcpip::Storage& ret = Dom::get(libsumo::TL_COMPLETE_DEFINITION_RYG, tlsID);
    std::vector<libsumo::TraCILogic> result;
    int numLogics = ret.readInt();
    while (numLogics-- > 0) {
        StoHelp::readCompound(ret, 5);
        libsumo::TraCILogic logic;
        logic.programID = StoHelp::readTypedString(ret);
        logic.type = StoHelp::readTypedInt(ret);
        logic.currentPhaseIndex = StoHelp::readTypedInt(ret);
        int numPhases = StoHelp::readCompound(ret);
        while (numPhases-- > 0) {
            StoHelp::readCompound(ret, 6);
            libsumo::TraCIPhase* phase = new libsumo::TraCIPhase();
            phase->duration = StoHelp::readTypedDouble(ret);
            phase->state = StoHelp::readTypedString(ret);
            phase->minDur = StoHelp::readTypedDouble(ret);
            phase->maxDur = StoHelp::readTypedDouble(ret);
            int numNext = StoHelp::readCompound(ret);
            while (numNext-- > 0) {
                phase->next.push_back(StoHelp::readTypedInt(ret));
            }
            phase->name = StoHelp::readTypedString(ret);
            logic.phases.emplace_back(phase);
        }
        int numParams = StoHelp::readCompound(ret);
        while (numParams-- > 0) {
            const std::vector<std::string> key_value = StoHelp::readTypedStringList(ret);
            logic.subParameter[key_value[0]] = key_value[1];
        }
        result.emplace_back(logic);
    }
    return result;
}


std::vector<std::string>
TrafficLight::getControlledJunctions(const std::string& tlsID) {
    return Dom::getStringVector(libsumo::TL_CONTROLLED_JUNCTIONS, tlsID);
}


std::vector<std::string>
TrafficLight::getControlledLanes(const std::string& tlsID) {
    return Dom::getStringVector(libsumo::TL_CONTROLLED_LANES, tlsID);
}


std::vector<std::vector<libsumo::TraCILink> >
TrafficLight::getControlledLinks(const std::string& tlsID) {
    tcpip::Storage& ret = Dom::get(libsumo::TL_CONTROLLED_LINKS, tlsID);
    std::vector< std::vector<libsumo::TraCILink> > result;
    ret.readInt();
    int numSignals = StoHelp::readTypedInt(ret);
    while (numSignals-- > 0) {
        std::vector<libsumo::TraCILink> controlledLinks;
        int numLinks = StoHelp::readTypedInt(ret);
        while (numLinks-- > 0) {
            std::vector<std::string> link = StoHelp::readTypedStringList(ret);
            controlledLinks.emplace_back(link[0], link[2], link[1]);
        }
        result.emplace_back(controlledLinks);
    }
    return result;
}


std::string
TrafficLight::getProgram(const std::string& tlsID) {
    return Dom::getString(libsumo::TL_CURRENT_PROGRAM, tlsID);
}


int
TrafficLight::getPhase(const std::string& tlsID) {
    return Dom::getInt(libsumo::TL_CURRENT_PHASE, tlsID);
}


std::string
TrafficLight::getPhaseName(const std::string& tlsID) {
    return Dom::getString(libsumo::VAR_NAME, tlsID);
}


double
TrafficLight::getPhaseDuration(const std::string& tlsID) {
    return Dom::getDouble(libsumo::TL_PHASE_DURATION, tlsID);
}


double
TrafficLight::getNextSwitch(const std::string& tlsID) {
    return Dom::getDouble(libsumo::TL_NEXT_SWITCH, tlsID);
}

int
TrafficLight::getServedPersonCount(const std::string& tlsID, int index) {
    tcpip::Storage content;
    content.writeUnsignedByte(libsumo::TYPE_INTEGER);
    content.writeInt(index);
    return Dom::getInt(libsumo::VAR_PERSON_NUMBER, tlsID, &content);
}

std::vector<std::string>
TrafficLight::getBlockingVehicles(const std::string& tlsID, int linkIndex) {
    tcpip::Storage content;
    content.writeUnsignedByte(libsumo::TYPE_INTEGER);
    content.writeInt(linkIndex);
    return Dom::getStringVector(libsumo::TL_BLOCKING_VEHICLES, tlsID, &content);
}

std::vector<std::string>
TrafficLight::getRivalVehicles(const std::string& tlsID, int linkIndex) {
    tcpip::Storage content;
    content.writeUnsignedByte(libsumo::TYPE_INTEGER);
    content.writeInt(linkIndex);
    return Dom::getStringVector(libsumo::TL_RIVAL_VEHICLES, tlsID, &content);
}

std::vector<std::string>
TrafficLight::getPriorityVehicles(const std::string& tlsID, int linkIndex) {
    tcpip::Storage content;
    content.writeUnsignedByte(libsumo::TYPE_INTEGER);
    content.writeInt(linkIndex);
    return Dom::getStringVector(libsumo::TL_PRIORITY_VEHICLES, tlsID, &content);
}

std::vector<libsumo::TraCISignalConstraint>
TrafficLight::getConstraints(const std::string& tlsID, const std::string& tripId) {
    std::vector<libsumo::TraCISignalConstraint> result;
    tcpip::Storage content;
    StoHelp::writeTypedString(content, tripId);
    tcpip::Storage& ret = Dom::get(libsumo::TL_CONSTRAINT, tlsID, &content);
    ret.readInt(); // components
    // number of items
    ret.readUnsignedByte();
    const int n = ret.readInt();
    for (int i = 0; i < n; ++i) {
        libsumo::TraCISignalConstraint c;
        c.signalId = StoHelp::readTypedString(ret);
        c.tripId = StoHelp::readTypedString(ret);
        c.foeId = StoHelp::readTypedString(ret);
        c.foeSignal = StoHelp::readTypedString(ret);
        c.limit = StoHelp::readTypedInt(ret);
        c.type = StoHelp::readTypedInt(ret);
        c.mustWait = StoHelp::readTypedByte(ret) != 0;
        result.push_back(c);
    }
    return result;
}

std::vector<libsumo::TraCISignalConstraint>
TrafficLight::getConstraintsByFoe(const std::string& foeSignal, const std::string& foeId) {
    std::vector<libsumo::TraCISignalConstraint> result;
    tcpip::Storage content;
    StoHelp::writeTypedString(content, foeId);
    tcpip::Storage& ret = Dom::get(libsumo::TL_CONSTRAINT_BYFOE, foeSignal, &content);
    ret.readInt(); // components
    // number of items
    ret.readUnsignedByte();
    const int n = ret.readInt();
    for (int i = 0; i < n; ++i) {
        libsumo::TraCISignalConstraint c;
        c.signalId = StoHelp::readTypedString(ret);
        c.tripId = StoHelp::readTypedString(ret);
        c.foeId = StoHelp::readTypedString(ret);
        c.foeSignal = StoHelp::readTypedString(ret);
        c.limit = StoHelp::readTypedInt(ret);
        c.type = StoHelp::readTypedInt(ret);
        c.mustWait = StoHelp::readTypedByte(ret) != 0;
        result.push_back(c);
    }
    return result;
}

LIBTRACI_PARAMETER_IMPLEMENTATION(TrafficLight, TL)

void
TrafficLight::setRedYellowGreenState(const std::string& tlsID, const std::string& state) {
    Dom::setString(libsumo::TL_RED_YELLOW_GREEN_STATE, tlsID, state);
}


void
TrafficLight::setPhase(const std::string& tlsID, const int index) {
    Dom::setInt(libsumo::TL_PHASE_INDEX, tlsID, index);
}


void
TrafficLight::setPhaseName(const std::string& tlsID, const std::string& name) {
    Dom::setString(libsumo::VAR_NAME, tlsID, name);
}


void
TrafficLight::setProgram(const std::string& tlsID, const std::string& programID) {
    Dom::setString(libsumo::TL_PROGRAM, tlsID, programID);
}


void
TrafficLight::setPhaseDuration(const std::string& tlsID, const double phaseDuration) {
    Dom::setDouble(libsumo::TL_PHASE_DURATION, tlsID, phaseDuration);
}


void
TrafficLight::setProgramLogic(const std::string& tlsID, const libsumo::TraCILogic& logic) {
    tcpip::Storage content;
    StoHelp::writeCompound(content, 5);
    StoHelp::writeTypedString(content, logic.programID);
    StoHelp::writeTypedInt(content, logic.type);
    StoHelp::writeTypedInt(content, logic.currentPhaseIndex);
    StoHelp::writeCompound(content, (int)logic.phases.size());
    for (const std::shared_ptr<libsumo::TraCIPhase>& phase : logic.phases) {
        StoHelp::writeCompound(content, 6);
        StoHelp::writeTypedDouble(content, phase->duration);
        StoHelp::writeTypedString(content, phase->state);
        StoHelp::writeTypedDouble(content, phase->minDur);
        StoHelp::writeTypedDouble(content, phase->maxDur);
        StoHelp::writeCompound(content, (int)phase->next.size());
        for (int n : phase->next) {
            StoHelp::writeTypedInt(content, n);
        }
        StoHelp::writeTypedString(content, phase->name);
    }
    StoHelp::writeCompound(content, (int)logic.subParameter.size());
    for (const auto& key_value : logic.subParameter) {
        StoHelp::writeTypedStringList(content, std::vector<std::string> {key_value.first, key_value.second});
    }
    Dom::set(libsumo::TL_COMPLETE_PROGRAM_RYG, tlsID, &content);
}


std::vector<libsumo::TraCISignalConstraint>
TrafficLight::swapConstraints(const std::string& tlsID, const std::string& tripId, const std::string& foeSignal, const std::string& foeId) {
    std::vector<libsumo::TraCISignalConstraint> result;
    tcpip::Storage content;
    content.writeByte(libsumo::TYPE_COMPOUND);
    content.writeInt(3);
    StoHelp::writeTypedString(content, tripId);
    StoHelp::writeTypedString(content, foeSignal);
    StoHelp::writeTypedString(content, foeId);
    tcpip::Storage& ret = Dom::get(libsumo::TL_CONSTRAINT_SWAP, tlsID, &content);
    ret.readInt(); // components
    // number of items
    ret.readUnsignedByte();
    const int n = ret.readInt();
    for (int i = 0; i < n; ++i) {
        libsumo::TraCISignalConstraint c;
        c.signalId = StoHelp::readTypedString(ret);
        c.tripId = StoHelp::readTypedString(ret);
        c.foeId = StoHelp::readTypedString(ret);
        c.foeSignal = StoHelp::readTypedString(ret);
        c.limit = StoHelp::readTypedInt(ret);
        c.type = StoHelp::readTypedInt(ret);
        c.mustWait = StoHelp::readTypedByte(ret) != 0;
        result.push_back(c);
    }
    return result;
}


void
TrafficLight::removeConstraints(const std::string& tlsID, const std::string& tripId, const std::string& foeSignal, const std::string& foeId) {
    tcpip::Storage content;
    content.writeByte(libsumo::TYPE_COMPOUND);
    content.writeInt(3);
    StoHelp::writeTypedString(content, tripId);
    StoHelp::writeTypedString(content, foeSignal);
    StoHelp::writeTypedString(content, foeId);
    Dom::set(libsumo::TL_CONSTRAINT_REMOVE, tlsID, &content);
}

std::string
to_string(const std::vector<double>& value) {
    std::ostringstream tmp;
    for (double d : value) {
        tmp << d << " ";
    }
    std::string tmp2 = tmp.str();
    tmp2.pop_back();
    return tmp2;
}


void
TrafficLight::setNemaSplits(const std::string& tlsID, const std::vector<double>& splits) {
    setParameter(tlsID, "NEMA.splits", to_string(splits));
}

void
TrafficLight::setNemaMaxGreens(const std::string& tlsID, const std::vector<double>& maxGreens) {
    setParameter(tlsID, "NEMA.maxGreens", to_string(maxGreens));
}

void
TrafficLight::setNemaCycleLength(const std::string& tlsID, double cycleLength) {
    setParameter(tlsID, "NEMA.cycleLength", std::to_string(cycleLength));
}

void
TrafficLight::setNemaOffset(const std::string& tlsID, double offset) {
    setParameter(tlsID, "NEMA.offset", std::to_string(offset));
}


LIBTRACI_SUBSCRIPTION_IMPLEMENTATION(TrafficLight, TL)

}


/****************************************************************************/

File: tools/traci/_trafficlight.py
# -*- coding: utf-8 -*-
# Eclipse SUMO, Simulation of Urban MObility; see https://eclipse.org/sumo
# Copyright (C) 2011-2023 German Aerospace Center (DLR) and others.
# This program and the accompanying materials are made available under the
# terms of the Eclipse Public License 2.0 which is available at
# https://www.eclipse.org/legal/epl-2.0/
# This Source Code may also be made available under the following Secondary
# Licenses when the conditions for such availability set forth in the Eclipse
# Public License 2.0 are satisfied: GNU General Public License, version 2
# or later which is available at
# https://www.gnu.org/licenses/old-licenses/gpl-2.0-standalone.html
# SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-or-later

# @file    _trafficlight.py
# @author  Michael Behrisch
# @date    2011-03-16

from __future__ import absolute_import
from .domain import Domain
from . import constants as tc
from .exceptions import TraCIException, deprecated


class Phase:

    def __init__(self, duration, state, minDur=-1, maxDur=-1, next=tuple(), name=""):
        self.duration = duration
        self.state = state
        self.minDur = minDur  # minimum duration (only for actuated tls)
        self.maxDur = maxDur  # maximum duration (only for actuated tls)
        self.next = next
        self.name = name

    def __repr__(self):
        name = "" if self.name == "" else ", name='%s'" % self.name
        return ("Phase(duration=%s, state='%s', minDur=%s, maxDur=%s, next=%s%s)" %
                (self.duration, self.state, self.minDur, self.maxDur, self.next, name))


class Logic:

    def __init__(self, programID, type, currentPhaseIndex, phases=None, subParameter=None):
        self.programID = programID
        self.type = type
        self.currentPhaseIndex = currentPhaseIndex
        self.phases = phases if phases is not None else []
        self.subParameter = subParameter if subParameter is not None else {}

    def getPhases(self):
        return self.phases

    def getSubID(self):
        return self.programID

    def getType(self):
        return self.type

    def getParameters(self):
        return self.subParameter

    def getParameter(self, key, default=None):
        return self.subParameter.get(key, default)

    def __repr__(self):
        return ("Logic(programID='%s', type=%s, currentPhaseIndex=%s, phases=%s, subParameter=%s)" %
                (self.programID, self.type, self.currentPhaseIndex, self.phases, self.subParameter))


def _readLogics(result):
    numLogics = result.readInt()
    logics = []
    for _ in range(numLogics):
        result.readCompound(5)
        programID = result.readTypedString()
        type = result.readTypedInt()
        currentPhaseIndex = result.readTypedInt()
        numPhases = result.readCompound()
        phases = []
        for __ in range(numPhases):
            result.readCompound(6)
            duration = result.readTypedDouble()
            state = result.readTypedString()
            minDur = result.readTypedDouble()
            maxDur = result.readTypedDouble()
            numNext = result.readCompound()
            next = tuple([result.readTypedInt() for ___ in range(numNext)])
            name = result.readTypedString()
            phases.append(Phase(duration, state, minDur, maxDur, next, name))
        logic = Logic(programID, type, currentPhaseIndex, tuple(phases))
        numParams = result.readCompound()
        for __ in range(numParams):
            key, value = result.readTypedStringList()
            logic.subParameter[key] = value
        logics.append(logic)
    return tuple(logics)


class Constraint:
    def __init__(self, signalId, tripId, foeId, foeSignal, limit, type, mustWait, active=True, param={}):
        self.signalId = signalId
        self.tripId = tripId
        self.foeId = foeId
        self.foeSignal = foeSignal
        self.limit = limit
        self.type = type
        self.mustWait = mustWait
        self.active = active
        self.param = param

    def getParameters(self):
        return self.param

    def getParameter(self, key, default=None):
        return self.param.get(key, default)

    def __repr__(self):
        param = ""
        if self.param:
            kvs = ["%s=%s" % (k, self.param[k]) for k in sorted(self.param.keys())]
            param = " param=%s" % '|'.join(kvs)
        return ("Constraint(signalId=%s tripId=%s, foeId=%s, foeSignal=%s, limit=%s, type=%s, mustWait=%s, active=%s%s)" %  # noqa
                (self.signalId, self.tripId, self.foeId, self.foeSignal,
                 self.limit, self.type, self.mustWait, self.active, param))


def _readLinks(result):
    result.readLength()
    numSignals = result.readInt()  # Length
    signals = []
    for _ in range(numSignals):
        # Type of Number of Controlled Links
        result.read("!B")
        # Number of Controlled Links
        nbControlledLinks = result.read("!i")[0]
        controlledLinks = []
        for __ in range(nbControlledLinks):
            result.read("!B")                       # Type of Link j
            link = result.readStringList()          # Link j
            controlledLinks.append(link)
        signals.append(controlledLinks)
    return signals


def _readConstraints(result):
    result.readLength()
    num = result.readInt()  # Length
    constraints = []
    for _ in range(num):
        signalId = result.readTypedString()
        tripId = result.readTypedString()
        foeId = result.readTypedString()
        foeSignal = result.readTypedString()
        limit = result.readTypedInt()
        type = result.readTypedInt()
        mustWait = bool(result.readTypedByte())
        active = bool(result.readTypedByte())
        paramItems = result.readTypedStringList()
        param = {}
        for i in range(0, len(paramItems), 2):
            param[paramItems[i]] = paramItems[i + 1]
        constraints.append(Constraint(signalId, tripId, foeId, foeSignal, limit, type, mustWait, active, param))
    return constraints


_RETURN_VALUE_FUNC = {tc.TL_COMPLETE_DEFINITION_RYG: _readLogics,
                      tc.TL_CONTROLLED_LINKS: _readLinks,
                      tc.TL_CONSTRAINT: _readConstraints,
                      tc.TL_CONSTRAINT_BYFOE: _readConstraints,
                      tc.TL_CONSTRAINT_SWAP: _readConstraints,
                      }


class TrafficLightDomain(Domain):

    Phase = Phase
    Logic = Logic

    def __init__(self, name="trafficlight", deprecatedFor=None):
        Domain.__init__(self, name, tc.CMD_GET_TL_VARIABLE, tc.CMD_SET_TL_VARIABLE,
                        tc.CMD_SUBSCRIBE_TL_VARIABLE, tc.RESPONSE_SUBSCRIBE_TL_VARIABLE,
                        tc.CMD_SUBSCRIBE_TL_CONTEXT, tc.RESPONSE_SUBSCRIBE_TL_CONTEXT,
                        _RETURN_VALUE_FUNC, deprecatedFor)

    def getRedYellowGreenState(self, tlsID):
        """getRedYellowGreenState(string) -> string

        Returns the named tl's state as a tuple of light definitions from
        rugGyYoO, for red, yed-yellow, green, yellow, off, where lower case letters mean that the stream
        has to decelerate.
        """
        return self._getUniversal(tc.TL_RED_YELLOW_GREEN_STATE, tlsID)

    def getAllProgramLogics(self, tlsID):
        """getAllProgramLogics(string) -> list(Logic)

        Returns a list of Logic objects.
        Each Logic encodes a traffic light program for the given tlsID.
        """
        return self._getUniversal(tc.TL_COMPLETE_DEFINITION_RYG, tlsID)

    getCompleteRedYellowGreenDefinition = deprecated("getCompleteRedYellowGreenDefinition")(getAllProgramLogics)

    def getControlledLanes(self, tlsID):
        """getControlledLanes(string) -> c

        Returns the list of lanes which are controlled by the named traffic light.
        """
        return self._getUniversal(tc.TL_CONTROLLED_LANES, tlsID)

    def getControlledLinks(self, tlsID):
        """getControlledLinks(string) -> list(list(list(string)))

        Returns the links controlled by the traffic light, sorted by the signal index and described by giving
        the incoming, outgoing, and via lane.
        """
        return self._getUniversal(tc.TL_CONTROLLED_LINKS, tlsID)

    def getProgram(self, tlsID):
        """getProgram(string) -> string

        Returns the id of the current program.
        """
        return self._getUniversal(tc.TL_CURRENT_PROGRAM, tlsID)

    def getPhase(self, tlsID):
        """getPhase(string) -> integer

        Returns the index of the current phase within the list of all phases of
        the current program.
        """
        return self._getUniversal(tc.TL_CURRENT_PHASE, tlsID)

    def getPhaseName(self, tlsID):
        """getPhaseName(string) -> string
        Returns the name of the current phase.
        """
        return self._getUniversal(tc.VAR_NAME, tlsID)

    def getNextSwitch(self, tlsID):
        """getNextSwitch(string) -> double

        Returns the absolute simulation time at which the traffic light is
        schedule to switch to the next phase (in seconds).
        """
        return self._getUniversal(tc.TL_NEXT_SWITCH, tlsID)

    def getPhaseDuration(self, tlsID):
        """getPhaseDuration(string) -> double

        Returns the total duration of the current phase (in seconds). This value
        is not affected by the elapsed or remaining duration of the current phase.
        """
        return self._getUniversal(tc.TL_PHASE_DURATION, tlsID)

    def getServedPersonCount(self, tlsID, index):
        """getServedPersonCount(string, int) -> int
        Returns the number of persons that would be served in the given phase
        """
        return self._getUniversal(tc.VAR_PERSON_NUMBER, tlsID, "i", index)

    def getBlockingVehicles(self, tlsID, linkIndex):
        """getBlockingVehicles(string, int) -> int
        Returns the list of vehicles that are blocking the subsequent block for
        the given tls-linkIndex
        """
        return self._getUniversal(tc.TL_BLOCKING_VEHICLES, tlsID, "i", linkIndex)

    def getRivalVehicles(self, tlsID, linkIndex):
        """getRivalVehicles(string, int) -> int
        Returns the list of vehicles that also wish to enter the subsequent block for
        the given tls-linkIndex (regardless of priority)
        """
        return self._getUniversal(tc.TL_RIVAL_VEHICLES, tlsID, "i", linkIndex)

    def getPriorityVehicles(self, tlsID, linkIndex):
        """getPriorityVehicles(string, int) -> int
        Returns the list of vehicles that also wish to enter the subsequent block for
        the given tls-linkIndex (only those with higher priority)
        """
        return self._getUniversal(tc.TL_PRIORITY_VEHICLES, tlsID, "i", linkIndex)

    def getConstraints(self, tlsID, tripId=""):
        """getConstraints(string, string) -> list(Constraint)
        Returns the list of rail signal constraints for the given rail signal.
        If tripId is not "", only constraints with the given tripId are
        returned. Otherwise, all constraints are returned
        """
        return self._getUniversal(tc.TL_CONSTRAINT, tlsID, "s", tripId)

    def getConstraintsByFoe(self, foeSignal, foeId=""):
        """getConstraintsByFoe(string, string) -> list(Constraint)
        Returns the list of rail signal constraints that have the given rail
        signal id as their foeSignal.
        If foeId is not "", only constraints with the given foeId are
        returned. Otherwise, all constraints are returned
        """
        return self._getUniversal(tc.TL_CONSTRAINT_BYFOE, foeSignal, "s", foeId)

    def setRedYellowGreenState(self, tlsID, state):
        """setRedYellowGreenState(string, string) -> None

        Sets the named tl's state as a tuple of light definitions from
        rugGyYuoO, for red, red-yellow, green, yellow, off, where lower case letters mean that the stream has
        to decelerate.
        """
        self._setCmd(tc.TL_RED_YELLOW_GREEN_STATE, tlsID, "s", state)

    def setLinkState(self, tlsID, tlsLinkIndex, state):
        """setLinkState(string, string, int, string) -> None
        Sets the state for the given tls and link index. The state must be one
        of rRgGyYoOu for red, red-yellow, green, yellow, off, where lower case letters mean that the stream has
        to decelerate.
        The link index is shown in the GUI when setting the appropriate junction
        visualization option.
        """
        fullState = list(self.getRedYellowGreenState(tlsID))
        if tlsLinkIndex >= len(fullState):
            raise TraCIException("Invalid tlsLinkIndex %s for tls '%s' with maximum index %s." % (
                tlsLinkIndex, tlsID, len(fullState) - 1))
        else:
            fullState[tlsLinkIndex] = state
            self.setRedYellowGreenState(tlsID, ''.join(fullState))

    def setPhase(self, tlsID, index):
        """setPhase(string, integer) -> None

        Switches to the phase with the given index in the list of all phases for
        the current program.
        """
        self._setCmd(tc.TL_PHASE_INDEX, tlsID, "i", index)

    def setPhaseName(self, tlsID, name):
        """setPhaseName(string, string) -> None

        Sets the name of the current phase within the current program
        """
        self._setCmd(tc.VAR_NAME, tlsID, "s", name)

    def setProgram(self, tlsID, programID):
        """setProgram(string, string) -> None

        Switches to the program with the given programID. The program must have
        been loaded earlier. The special value 'off' can always be used to
        switch off the traffic light.
        """
        self._setCmd(tc.TL_PROGRAM, tlsID, "s", programID)

    def getNemaPhaseCalls(self, tlsID):
        """getNemaPhaseCalls(string) -> list(string)
        Get the vehicle calls for the phases.
        The output is vehicle calls (coming from the detectors) for the phases.
        """
        vehCallStr = self.getParameter(tlsID, "NEMA.phaseCall")
        return vehCallStr.split(",")

    def setNemaSplits(self, tlsID, splits):
        """setNemaSplits(string, list(string)) -> None
        Set a new set of splits to the given NEMA-controller.
        This function is only effective for NEMA type of controllers.
        The new splits will be implemented in the next cycle of the control.
        The value must be a space-separated list of 8 numbers with each number
        being the time in seconds for NEMA-phases 1 to 8.
        Time 0 must be used of the phase does not exists.
        Example: “11.0 34.0 15.0 20.0 11.0 34.0 15.0 20.0" (gives total cycle length of 80s)
        """
        if type(splits) == list:
            splits = ' '.join(map(str, splits))
        self.setParameter(tlsID, "NEMA.splits", splits)

    def setNemaMaxGreens(self, tlsID, maxGreens):
        """setNemaMaxGreens(string, list(string)) -> None
        Set a new set of splits to the given NEMA-controller.
        This function is only effective for NEMA type of controllers.
        The new max green will be implemented in the next cycle of the control.
        The value must be a space-separated list of 8 numbers with each number
        being the time in seconds for NEMA-phases 1 to 8.
        Time 0 must be used of the phase does not exists.
        Example: “11.0 34.0 15.0 20.0 11.0 34.0 15.0 20.0"
        """
        if type(maxGreens) == list:
            maxGreens = ' '.join(map(str, maxGreens))
        self.setParameter(tlsID, "NEMA.maxGreens", maxGreens)

    def setNemaCycleLength(self, tlsID, cycleLength):
        """setNemaCycleLength(string, double) -> None
        Set a new cycle length to the given NEMA-controller.
        This function is only effective for NEMA type of controllers.
        The new cycle length will be implemented in the next cycle of the control.
        This function should be used with setNemaSplits or setNemaMaxGreen.
        """
        self.setParameter(tlsID, "NEMA.cycleLength", str(cycleLength))

    def setNemaOffset(self, tlsID, offset):
        """setNemaOffset(string, double) -> None
        Set a new offset to the given NEMA-controller.
        This function is only effective for NEMA type controllers when operating under coordinated mode.
        The new offset will be implemented in the next cycle of the control by adjusting
        the actual green time of the coordinated phase.
        There is no transition implemented in the NEMA-controller for changing the offset.
        It’s expected that the users will control the change of the offset in each cycle
        to implement their own transition algorithm.
        """
        self.setParameter(tlsID, "NEMA.offset", str(offset))

    def setPhaseDuration(self, tlsID, phaseDuration):
        """setPhaseDuration(string, double) -> None

        Set the remaining phase duration of the current phase in seconds.
        This value has no effect on subsquent repetitions of this phase.
        """
        self._setCmd(tc.TL_PHASE_DURATION, tlsID, "d", phaseDuration)

    def setProgramLogic(self, tlsID, tls):
        """setProgramLogic(string, Logic) -> None

        Sets a new program for the given tlsID from a Logic object.
        See getCompleteRedYellowGreenDefinition.
        """
        format = "tsiit"
        values = [5, tls.programID, tls.type, tls.currentPhaseIndex, len(tls.phases)]
        for p in tls.phases:
            format += "tdsddt"
            values += [6, p.duration, p.state, p.minDur, p.maxDur, len(p.next)]
            for n in p.next:
                format += "i"
                values += [n]
            format += "s"
            values += [p.name]
        # subparams
        format += "t"
        values += [len(tls.subParameter)]
        for par in tls.subParameter.items():
            format += "l"
            values += [par]
        self._setCmd(tc.TL_COMPLETE_PROGRAM_RYG, tlsID, format, *values)

    setCompleteRedYellowGreenDefinition = deprecated("setCompleteRedYellowGreenDefinition")(setProgramLogic)

    def swapConstraints(self, tlsID, tripId, foeSignal, foeId):
        """swapConstraints(string, string, string, string) -> list(Constraint)
        Reverse the given constraint and return list of new constraints that
        were created (by swapping) to avoid deadlock.
        """
        return self._getUniversal(tc.TL_CONSTRAINT_SWAP, tlsID, "tsss", 3, tripId, foeSignal, foeId)

    def removeConstraints(self, tlsID, tripId, foeSignal, foeId):
        """removeConstraints(string, string, string, string)
        remove constraints with the given values. Any combination of inputs may
        be set to "" to act as a wildcard filter """
        self._setCmd(tc.TL_CONSTRAINT_REMOVE, tlsID, "tsss", 3, tripId, foeSignal, foeId)

    def updateConstraints(self, vehID, tripId=""):
        """getConstraints(string, string)
        Removes all constraints that can no longer be met because the route of
        vehID does not pass traffic light of the constraint with the given tripId.
        This includes constraints on tripId as well as constraints where tripId
        is the foeId.
        If tripId is "", the current tripId of vehID is used.
        """
        self._setCmd(tc.TL_CONSTRAINT_UPDATE, vehID, "s", tripId)
